ReactAndroid/src/main/java/com/facebook/react/views/scroll/MaintainVisibleScrollPositionHelper.java

```diff
@@ -0,0 +1,186 @@
+import com.facebook.react.bridge.ReactContext;
+import com.facebook.react.bridge.ReadableMap;
+import com.facebook.react.bridge.UIManager;
+import com.facebook.react.bridge.UIManagerListener;
+import com.facebook.react.bridge.UiThreadUtil;
+import com.facebook.react.uimanager.UIManagerHelper;
+import com.facebook.react.uimanager.common.ViewUtil;
+import com.facebook.react.views.view.ReactViewGroup;
+import com.facebook.react.views.scroll.ReactScrollViewHelper.HasSmoothScroll;
+
+import java.lang.ref.WeakReference;
+
+/**
+ * Manage state for the maintainVisibleContentPosition prop.
+ *
+ * This uses UIManager to listen to updates and capture position of items before and after layout.
+ */
+public class MaintainVisibleScrollPositionHelper<ScrollViewT extends ViewGroup & HasSmoothScroll> implements UIManagerListener {
+  private final ScrollViewT mScrollView;
+  private final boolean mHorizontal;
+  private @Nullable Config mConfig;
+  private @Nullable WeakReference<View> mFirstVisibleView = null;
+  private @Nullable Rect mPrevFirstVisibleFrame = null;
+  private boolean mListening = false;
+
+  public static class Config {
+    public final int minIndexForVisible;
+    public final @Nullable Integer autoScrollToTopThreshold;
+
+    Config(int minIndexForVisible, @Nullable Integer autoScrollToTopThreshold) {
+      this.minIndexForVisible = minIndexForVisible;
+      this.autoScrollToTopThreshold = autoScrollToTopThreshold;
+    }
+
+    static Config fromReadableMap(ReadableMap value) {
+      int minIndexForVisible = value.getInt("minIndexForVisible");
+      Integer autoScrollToTopThreshold =
+        value.hasKey("autoscrollToTopThreshold")
+          ? value.getInt("autoscrollToTopThreshold")
+          : null;
+      return new Config(minIndexForVisible, autoScrollToTopThreshold);
+    }
+  }
+
+  public MaintainVisibleScrollPositionHelper(ScrollViewT scrollView, boolean horizontal) {
+    mScrollView = scrollView;
+    mHorizontal = horizontal;
+  }
+
+  public void setConfig(@Nullable Config config) {
+    mConfig = config;
+  }
+
+  /**
+   * Start listening to view hierarchy updates. Should be called when this is created.
+   */
+  public void start() {
+    if (mListening) {
+      return;
+    }
+    mListening = true;
+    getUIManagerModule().addUIManagerEventListener(this);
+  }
+
+  /**
+   * Stop listening to view hierarchy updates. Should be called before this is destroyed.
+   */
+  public void stop() {
+    if (!mListening) {
+      return;
+    }
+    mListening = false;
+    getUIManagerModule().removeUIManagerEventListener(this);
+  }
+
+  /**
+   * Update the scroll position of the managed ScrollView. This should be called after layout
+   * has been updated.
+   */
+  public void updateScrollPosition() {
+    if (mConfig == null
+      || mFirstVisibleView == null
+      || mPrevFirstVisibleFrame == null) {
+      return;
+    }
+
+    View firstVisibleView = mFirstVisibleView.get();
```

Comment:
If willDispatchViewUpdates is called from ui thread (which I’m not sure if it ever does) the dispatch to ui thread will be done synchronously by runOnUIThread. Ideally we could have a way to observe before / after views are updated on the ui thread but wanted to avoid adding more methods to UIManagerObserver for old arch. In my tests this always worked so I was fine with this solution.

Reply:
@janicduplessis `firstVisibleView` may be null, i ran into.
`
E/unknown:ReactNative: Exception in native call
    java.lang.NullPointerException: Attempt to invoke virtual method 'void android.view.View.getHitRect(android.graphics.Rect)' on a null object reference
        at com.facebook.react.views.scroll.MaintainVisibleScrollPositionHelper.updateScrollPositionInternal(MaintainVisibleScrollPositionHelper.java:107)
        at com.facebook.react.views.scroll.MaintainVisibleScrollPositionHelper.updateScrollPosition(MaintainVisibleScrollPositionHelper.java:97)
        at com.facebook.react.views.scroll.ReactScrollView.onLayoutChange(ReactScrollView.java:1128)
        at android.view.View.layout(View.java:20717)
        at android.view.ViewGroup.layout(ViewGroup.java:6198)
        at com.facebook.react.uimanager.NativeViewHierarchyManager.updateLayout(NativeViewHierarchyManager.java:254)
        at com.facebook.react.uimanager.NativeViewHierarchyManager.updateLayout(NativeViewHierarchyManager.java:222)
`

Reply:
I see, yes it is missing a null check, I'll open a PR to fix it.